doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
```

Comment:
Also, Caffe2 is using `unique_ptr` in its Workspace [definition](https://github.com/caffe2/caffe2/blob/v0.7.0/caffe2/core/workspace.h#L56):

```cpp
typedef CaffeMap<string, unique_ptr<Blob> > BlobMap;
```

Why do we use `shared_ptr` here?

To my understand, `unique_ptr` is the right choice here because it transfers ownership of the pointed object when copy, whereas `shared_ptr` increment the reference count when copy.